File: Day-10/kotlin/unbekannt/day-10.kt
package unbekannt

import java.io.File

fun main() {
    val input = File("input-day10.txt").readLines()

    var X = 1
    var adds: Addx? = null
    val signals = mutableListOf<Int>()
    val lines = mutableListOf<String>()

    var signal = 1
    var index = 0
    var currentLine = ""
    while (index < input.size){
        var wasChanged = false
        val command = input[index]
        if (signal == 20 || (signal - 20) % 40 == 0) {
            signals.add(signal * X)
        }
        currentLine += if (signal % 40 in (X .. X+2) || (signal % 40 == 0 && 40 in (X .. X+2) )) "#" else "."

        if (currentLine.length == 40) {
            lines.add(currentLine)
            currentLine = ""
        }

        if(adds != null){
            if (adds.siganlIndex == signal) {
                X += adds.value
                adds = null
                wasChanged = true
            }
        }


        if (command.startsWith("addx") && !wasChanged && adds == null) {
            adds = command.split(" ").let { Addx(it.last().toInt(), signal + 1) }
        }
        signal++
        if(adds == null) index++
    }
    println(signals.sum())
    lines.forEach {
        println(it)
    }
}


data class Addx(val value: Int, val siganlIndex: Int)

File: Day-10/rust/Marie/day10.rs
#![feature(iter_array_chunks)]

use std::{fs, str::FromStr};

const TARGET_CYCLES: &[usize] = &[20, 60, 100, 140, 180, 220];

fn main() {
    let input = fs::read_to_string("inputs/day10.txt").unwrap();
    let registers: Vec<i32> = input
        .lines()
        .map(|line| Instruction::from_str(line).unwrap())
        .fold(vec![1], |mut cpu, instruction| {
            match instruction {
                Instruction::Addx(amount) => {
                    let x = *cpu.last().unwrap();
                    cpu.push(x);
                    cpu.push(x + amount);
                }
                Instruction::Noop => {
                    cpu.push(*cpu.last().unwrap());
                }
            }
            cpu
        });
    let sum: usize = registers
        .iter()
        .enumerate()
        .filter_map(|(index, value)| {
            if TARGET_CYCLES.iter().any(|v| *v == index + 1) {
                Some(*value as usize * (index + 1))
            } else {
                None
            }
        })
        .sum();
    println!("{sum}");

    (0..240)
        .into_iter()
        .map(|pixel| {
            let x = registers[pixel as usize];
            let sprite_pos = x % 40;
            let sprite_pixel = pixel % 40;
            if sprite_pos == sprite_pixel
                || sprite_pos - 1 == sprite_pixel
                || sprite_pos + 1 == sprite_pixel
            {
                '#'
            } else {
                '.'
            }
        })
        .array_chunks::<40>()
        .for_each(|line| println!("{}", line.iter().collect::<String>()))
}

enum Instruction {
    Addx(i32),
    Noop,
}

impl FromStr for Instruction {
    type Err = ();

    fn from_str(value: &str) -> Result<Self, Self::Err> {
        let mut line = value.split(' ');
        let instruction = match line.next().unwrap() {
            "addx" => {
                let amount = line.next().and_then(|a| a.parse::<i32>().ok()).unwrap();
                Instruction::Addx(amount)
            }
            "noop" => Instruction::Noop,
            _ => panic!("Invalid instruction"),
        };
        Ok(instruction)
    }
}
